File: singleton/java/Singleton.java
public class Singleton {

    private int counter = 0;

    private Singleton() {
        System.out.println("Inizialize Singleton ");
    }

    private static class SingletonHolder {
        private static final Singleton INSTANCE = new Singleton();
    }

    public static Singleton getInstance() {
        return SingletonHolder.INSTANCE;
    }

    public void doSomething() {
        counter++;
        System.out.println("this object: " + this);
        System.out.println("Identity HashCode: " + System.identityHashCode(this));
        System.out.println("From Thread: " + Thread.currentThread().getName());
        System.out.println("Call counter: " + counter);
    }
}

File: singleton/cpp/Singleton.cpp
#include "Singleton.h"
#include <iostream>

// Реализация getInstance (Meyers' Singleton)
Singleton& Singleton::getInstance() {
    static Singleton instance; // Создастся только один раз
    return instance;
}

// Реализация конструктора
Singleton::Singleton() {
    std::cout << ">>> Singleton: Конструктор вызван" << std::endl;
}

// Реализация деструктора
Singleton::~Singleton() {
    std::cout << ">>> Singleton: Деструктор вызван" << std::endl;
}

// Реализация бизнес-логики
void Singleton::doSomething() {
    std::cout << "Work done! Я нахожусь по адресу: " << this << std::endl;
}
